OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -257,46 +257,82 @@ bool TracerThread::OpenSampling(const std::vector<int32_t>& cpus) {
   return true;
 }
 
-bool TracerThread::InitGpuTracepointEventProcessor() {
-  int amdgpu_cs_ioctl_id = GetTracepointId("amdgpu", "amdgpu_cs_ioctl");
-  if (amdgpu_cs_ioctl_id == -1) {
-    return false;
+void TracerThread::OpenRingBuffersOrRedirectOnExisting(
+    const absl::flat_hash_map<int32_t, int>& fds_per_cpu,
+    absl::flat_hash_map<int32_t, int>* ring_buffer_fds_per_cpu,
+    std::vector<PerfEventRingBuffer>* ring_buffers,
+    uint64_t ring_buffer_size_kb, std::string_view buffer_name_prefix) {
+  // Redirect all events on the same cpu to a single ring buffer.
+  for (const auto& cpu_and_fd : fds_per_cpu) {
+    int32_t cpu = cpu_and_fd.first;
+    int fd = cpu_and_fd.second;
+    if (ring_buffer_fds_per_cpu->contains(cpu)) {
+      // Redirect to the already opened ring buffer.
+      int ring_bugger_fd = ring_buffer_fds_per_cpu->at(cpu);
+      perf_event_redirect(fd, ring_bugger_fd);
+    } else {
+      // Create a ring buffer for this cpu.
+      int ring_buffer_fd = fd;
+      std::string buffer_name =
+          absl::StrFormat("%s_%d", buffer_name_prefix, cpu);
+      ring_buffers->emplace_back(ring_buffer_fd, ring_buffer_size_kb,
+                                 buffer_name);
+      ring_buffer_fds_per_cpu->emplace(cpu, ring_buffer_fd);
+    }
   }
-  int amdgpu_sched_run_job_id =
-      GetTracepointId("amdgpu", "amdgpu_sched_run_job");
-  if (amdgpu_sched_run_job_id == -1) {
-    return false;
+}
+
+bool TracerThread::OpenRingBuffersForTracepoint(
```

Comment:
Maybe one final suggestion: One way of reducing the amount of arguments is to combine some of them into a struct, if that can be done in a meaningful way. It can also become more readable if you group arguments that belong together into a single structure. 

Feel free to ignore this, I don't think you would gain a lot here (if anything).

Reply:
Maybe `tracepoint_category`, `tracepoint_name` and `cpus` could be grouped, but the others not really as some are output parameters and others input/output parameters. So I'll leave it as it is as I don't think it would improve a lot.

Reply:
Agree.